test/gurka/test_route_incidents.cc

```diff
@@ -0,0 +1,685 @@
+      ASSERT_GE(i.begin_shape_index(), last_shape_index) << " leg incidents were out of order";
+      last_shape_index = i.begin_shape_index();
+    }
+    incident_count += leg.incidents_size();
+  }
+
+  // check all the incidents are in the right place in the route
+  for (const auto& i : locations) {
+    const auto& leg = api.trip().routes(0).legs(i.leg_index);
+    // find the incident
+    const valhalla::TripLeg::Incident* incident = nullptr;
+    for (int j = 0; j < leg.incidents_size(); ++j) {
+      if (leg.incidents(j).id() == i.id) {
+        incident = &leg.incidents(j);
+        break;
+      }
+    }
+    ASSERT_TRUE(incident != nullptr) << "Couldn't find incident " << i.id << " on leg " << i.leg_index
+                                     << " on edge " << i.begin_edge_index;
+
+    // check that we are on the right edges of the path
+    const auto& begin_edge = leg.node(i.begin_edge_index).edge();
+    ASSERT_TRUE(begin_edge.begin_shape_index() <= incident->begin_shape_index() &&
+                incident->begin_shape_index() <= begin_edge.end_shape_index())
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.begin_edge_index
+        << " begins on wrong edge";
+    const auto& end_edge = leg.node(i.end_edge_index).edge();
+    ASSERT_TRUE(end_edge.begin_shape_index() <= incident->end_shape_index() &&
+                incident->end_shape_index() <= end_edge.end_shape_index())
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.end_edge_index
+        << " ends on wrong edge";
+
+    // check that we are on the right shape points
+    const auto& lengths = leg_lengths[i.leg_index];
+    auto begin_edge_length =
+        lengths[begin_edge.end_shape_index()] - lengths[begin_edge.begin_shape_index()];
+    auto begin_edge_pct =
+        (lengths[incident->begin_shape_index()] - lengths[begin_edge.begin_shape_index()]) /
+        begin_edge_length;
+    ASSERT_NEAR(begin_edge_pct, i.begin_pct, .01)
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.begin_edge_index
+        << " begins at wrong location along the edge";
+    auto end_edge_length =
+        lengths[end_edge.end_shape_index()] - lengths[end_edge.begin_shape_index()];
+    auto end_edge_pct =
+        (lengths[incident->end_shape_index()] - lengths[end_edge.begin_shape_index()]) /
+        end_edge_length;
+    ASSERT_NEAR(end_edge_pct, i.end_pct, .1)
+        << "Incident " << i.id << " on leg " << i.leg_index << " on edge " << i.end_edge_index
+        << " ends at wrong location along the edge";
+  }
+
+  ASSERT_EQ(incident_count, locations.size()) << "Expected number of incidents does not match actual";
+}
+
+// provides us an easy way to mock having incident tiles, each test can override the tile in question
+// a bit more work is needed if we want to do it for more than one tile at a time
+struct test_reader : public baldr::GraphReader {
```

Comment:
this is what makes it possible to actually test this without incident data. we make a testable graph reader that allows  us to directly modify the incident tile that it uses to answer the GetIncidents function when triplegbuilder calls it. at the moment it only has one tile. which makes it not quite good enough, somehow though i got lucky and the edge indices within the tiles didnt collide. i can expand this to support multiple tiles at a time though to remove that issue if reviewers feel that it is worth it

Reply:
(i did expand it to support multiple tiles at a time in the latest commit so its not an issue anymore)